### ISSUE TITLE
Alternative encoding of URL key and id and URL length

### ISSUE THREAD

**Issue by elrido** [PrivateBin/PrivateBin#638]
**TL;DR:** I compared some alternative text encodings for the paste key and ID and after discussing the pros and cons may consider to implement such a change as a configurable, non-default option for those that like shorter, weirder looking URLs. [Play with the fiddle](https://jsfiddle.net/jfntqmr0/4/) and generate new, random URLs by clicking on "Run" on the top left and please share your thoughts.

**Background:** Both the frequent questions on how to use the URL-shortener feature, as well as it's inception or the requests for producing shorter URLs in the first place #626 show that a number of users would prefer shorter URLs, even trading in security for it.

I recently stumbled on [ecoji](https://github.com/keith-turner/ecoji), which is a base1024 encoding (every 8 bits get encoded to 10 bits vs the 6 bits in base64), using a [list of 1024 sequential, moderated emojis](https://github.com/keith-turner/ecoji/blob/master/mapping.go#L4) plus some extra ones for padding. These would obviously make for visually quite distinct URLs on their own, thought at this time there is no browser compatible JS implementation (for the fiddle I had to set up a [little API service using the PHP implementation](https://github.com/elrido/ecoji-php-api)).

The main take-away for me from this concept is that a base above 256 lets you use less (unicode) symbols to represent more bytes of data. On the other hand, in the browsers address bar, these URLs get usually displayed with sans serif fonts, while emojis are wider then most lower case latin characters. We could potentially generate shorter (looking) URLs, without reducing key size (= safety) or reducing the amount of possible pastes.

So I wondered if there are higher bases using larger unicode character sets. Also, at base 65536 the 16 bits would align with our binary data lengths, making padding unnecessary (ID and key are always 64 bits and 256 bits long, respectively). Sure enough, there are already JS implementations for other bases: [2048](https://github.com/qntm/base2048), [32768](https://github.com/qntm/base32768) and
[65536](https://github.com/qntm/base65536).

**Caveats:** (that I can think of so far)
- Not everyone will like the look of these URLs - hence I wouldn't make this a default, rather a configuration option.
- While we understand the issues with ASCII-only GET-parameters and hashes, introducing unicode characters into these could cause many unpredictable side-effects, for example certain filters in mail servers, chat systems or on websites might break these or refuse to accept these URLs as valid links. I would imagine that our own `urls2links()` function or the markdown parser used on pasted texts will likely not detect these, for example.
- To me, the ecoji URLs look visually pleasing and would make it easier to distinguish to different paste URLs to us humans. Unfortunately they are even wider then the current base58 ones. base2048 and base32768 tend to use less horizontal space.
- One would naively think that the encoding with the largest base and that happens to align with our binary data size, making padding unnecessary, would produce the shortest looking URLs, but as the fiddle shows, that is often not the case. The length varies also significantly depending on the data, be sure to re-"Run" the fiddle a few times to get some idea of this - occasionally base2048 produces the shortest URL, often base32768 is shorter then base65536.
- You will notice missing characters in the base65536 URLs - That encoding uses a mix of characters from different writing systems that are not found in a single font. TrueType and OpenType (and WOFF, which is a wrapper format for these) file formats use a 16 bit ID for the symbols described by them, hence we still have to split fonts into families to cover different groups of unicode characters. See for example the Android default [Noto font family](https://www.google.com/get/noto/). You can totally mix such different writing systems in a single document, but will have to potentially use different fonts for different sections, to display all of the symbols correctly. But without applying different fonts per symbol, some writing systems can't easily be mixed inside a single HTML DOM element, for example or you'd have to add `<span>`s per font.
- In the fiddle I included byte counts, as the unicode characters obviously tend to use 16 or 32 bits compared to the 8 bits for hex, base64 or base58 encoded pieces.
- The vertical alignment of the different symbols on all of these isn't optimal, but that entirely depends on the font that this gets displayed in on the platform they are shared on.
- Changing the key format could be handled backwards compatible, at least if we only have to consider base58, ecoji (base1024), base2048 or base65536 as they seem to use different codepoint ranges. It should be possible to select the appropriate decoder based on that. Or we just attempt different decodings until we get a match.
- Changing the ID format would require the support of that encoding on the server side. That complication was why I ended up avoiding changing the ID from hex to base58, [contrary to my earlier intention](https://github.com/PrivateBin/PrivateBin/pull/428#issuecomment-490515528).

**Comment by rugk** [comment]
First of all, I like the general idea of using emojis and so on here. :blush: 

That said, I think the reason behind feature requests like https://github.com/PrivateBin/PrivateBin/issues/626 (/cc'ing @tantoinet) was to have URLs that you can e.g. say over the phone etc.

So if we consider this **use case**: (say URL via voice)
* all the Chinese etc. characters are not a good idea, given that a huge population of the world cannot spell – let alone – write it on their computer keyboard
* emojis, on the other hand, are a good idea, given they can likely be spelled quite good and most OS'es can nowadays write them
  * only problem: Unicode standards – we'd better not use that latest of all Unicode standards then, as not each OS may support it yet
* the visual width does not matter

So in general I like the idea.

## Concrete implementation

Your "ecoji for key" example, however, is not really better than ASCII chars (base58)…
Okay, no, I just need to ignore the char counting you implemented, as Unicode emojis often use 2 chars to represent 1 emoji.

base58: 44 chars for key
emoji: 29 chars, if I counted correctly (visually)

So an improvement yeah.

However, 1024 emojis seems a little few. And given their comments there, it's not really "curated", but rather autogenerated from Emoji v11.
If we'd take [skin tone modifiers](https://emojipedia.org/emoji-modifier-sequence/) into account we likely increase the number even more.
Also these family emojis e.g. Basically [all these that require a modifier](https://emojipedia.org/emoji-zwj-sequence/).
The problem here may be that not all are "typebable" on all devices.
AFAIK skin tones should be, that's not a knew feature. When Unicode introduced the same for family/relationship emojis (like these holding hands, kissing etc.) most vendors gave up to support that _together_ with each possible skin tone modifier.
[Microsoft does](https://blog.emojipedia.org/diverse-emoji-families-come-to-windows/) e.g. – for displaying, but not for input – and thus has 52.000 emojis! Wow… that's a lot of entropy we had there, so we could likely get the length down to 10 emojis or so if we'd support that all, hehe.
(I know, we could likely calculate the length mathematically, but I'm just too lazy to do that now.)
The input problem remains, but well…

Also, the list seems to exclude the flags – which are [one of the biggest categories](https://www.youtube.com/watch?v=lIFE7h3m40U). Though yeah, yet again people may not know a country and type it.

**Comment by elrido** [comment]
The use case of repeating the URL verbally was not on my radar at all and I can't imagine that is a very practical way of doing it. The number of characters or symbols must lead to a number of typos and a single typo would cause the paste to fail to decrypt. But yeah, that is a strong argument in favour of shorter URLs.

I though about the UI a bit and since we do support more then one URL types for decryption anyway (old base64 and base58 as of version 1.3, at this time), we could maybe change the layout of the "paste success" message box and include all different forms of the URL for the user to pick what they like (I would still make the new URL encoding an optional feature, but if enabled it would add one to the list):

> Your paste is ready, choose one of the following ways to share it:
> - https://example.com/?c240fc94493e90f3#Dt5iveiu93MY6L3kPXckBGLTD5VyTi2xLZEStaxXB5yU ["Copy URL" button]
> - https://example.com/?🚅🆙🚡🌏🏇🏝🎈☕#🙈📵🏤🚳💺🤸🎮🙎🛅💬🔕😎🎂🚬🆒🍒🦝🥠🇳💗🐳📖💼😲💄🗻☕☕ ["Copy URL" button]
> - ["QR code" button]
> - ["Email"-button]
> 
> Use the following link to delete your paste: Delete paste

The idea to incorporate composed unicode characters is a good one, even if it means we would have to create our own mappings and libraries (for js and php). It would let us use a large base 65536 that fits our data sizes, avoiding wasted padding space and we don't care if the URL becomes longer in bytes, if it is shorter in display length.

We probably should try to avoid some symbols, especially those not "Recommended for General Interchange". I can also imagine that ecoji avoided the flags as some are not recognized by all governements, like 🇹🇼 and 🏴󠁣󠁮󠀵󠀴󠁿 (the latter isn't RGI) or the European example of 🏴󠁧󠁢󠁥󠁮󠁧󠁿, 🏴󠁧󠁢󠁳󠁣󠁴󠁿 and 🏴󠁧󠁢󠁷󠁬󠁳󠁿. We might get away with it, as we use them in a non-political context. But it would be possible that a sequence gets generated that might be interpreted as a slur or offensive to some.

Anyway, I suggest I create a proof of concept of a JS library first, but in a separate project, then report back here when I could update the JS fiddle to make use of it.

**Comment by rugk** [comment]
I guess a library would be useful indeed. Actually, the difficult part of course is just choosing and generating the emoji maps.

I think the best architecture might just be a script/parser that parses an Unicode list of all emojis (so we can easily update it later), and removes certain (types of) emojis based on some criteria that we can edit.


But maybe we can also look how other's do that. E.g. I know Riot also uses emojis for verifying the security, so they must have a mapping there, too.

**Comment by cburgess6969** [comment]
Not all browsers and operting systems support the easy input and rendering of emojis, and unlike base58 which can be written on paper, it would be hard(er) to reliably write emojis on paper. So I think it would be fine to primarily use base58, then convert to emoji as a "vanity URL".

**Comment by rugk** [comment]
Thowing in some things i found:
https://hashids.org/
https://github.com/ai/nanoid (the security section sounds good and relevant for US, especially that mod was not a uniform aka safe operation there is new to me)

**Comment by elrido** [comment]
So this was a fun idea, but I think we collected enough reasons that speak against it and I've abandoned the project as well, so I'll close this a wont-fix.
